zng/fuser/fuser.go

```diff
@@ -0,0 +1,173 @@
+// Packaage fuser transforms records to a unified schema.
+package fuser
+
+import (
+	"errors"
+
+	"github.com/brimsec/zq/proc/spill"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+	"github.com/brimsec/zq/zng/resolver"
+)
+
+// Fuser buffers records written to it, assembling from them a unified schema of
+// fields and types.  Fuser then transforms those records to the unified schema
+// as they are read back from it.
+type Fuser struct {
+	zctx        *resolver.Context
+	memMaxBytes int
+
+	nbytes   int
+	recs     []*zng.Record
+	slotByID [][]int
+	slots    []slot
+	spiller  *spill.File
+	types    resolver.Slice
+	uberType *zng.TypeRecord
+}
+
+type slot struct {
+	zv        zcode.Bytes
+	container bool
+}
+
+// New returns a new Fuser.  The Fuser buffers records in memory until their
+// cumulative size (measured in zcode.Bytes length) exceeds memMaxBytes, at
+// which point it buffers them in a temporary file.
+func New(zctx *resolver.Context, memMaxBytes int) *Fuser {
+	return &Fuser{zctx: zctx, memMaxBytes: memMaxBytes}
+}
+
+// Close removes the receiver's temporary file if it created one.
+func (f *Fuser) Close() error {
+	if f.spiller != nil {
+		return f.spiller.CloseAndRemove()
+	}
+	return nil
+}
+
+// Write buffers rec. If called after Read, Write panics.
+func (f *Fuser) Write(rec *zng.Record) error {
+	if f.finished() {
+		panic("fuser: write after read")
+	}
+	id := rec.Type.ID()
+	typ := f.types.Lookup(id)
+	if typ == nil {
+		f.types.Enter(id, rec.Type)
+	}
+	if f.spiller != nil {
+		return f.spiller.Write(rec)
+	}
+	return f.stash(rec)
+}
+
+func (f *Fuser) stash(rec *zng.Record) error {
+	f.nbytes += len(rec.Raw)
+	if f.nbytes >= f.memMaxBytes {
+		var err error
+		f.spiller, err = spill.NewTempFile()
+		if err != nil {
+			return err
+		}
+		if err := f.spiller.Write(rec); err != nil {
+			return err
+		}
```

Comment:
Shouldn't this go after the recs iteration to keep the order or does this not matter?

Reply:
I think fuse should preserve order and created #1904 to track that.